File: test/UserTest.java
import com.google.appengine.api.datastore.DatastoreService;
import com.google.appengine.api.datastore.DatastoreServiceFactory;
import com.google.appengine.api.datastore.Query;
import org.junit.*;
import play.test.*;
import models.*;

import com.google.appengine.tools.development.testing.LocalDatastoreServiceTestConfig;
import com.google.appengine.tools.development.testing.LocalServiceTestHelper;
import java.util.Date;

public class UserTest extends UnitTest {

    private final LocalServiceTestHelper helper =
        new LocalServiceTestHelper(new LocalDatastoreServiceTestConfig());

    @Before
    public void setUp() {
        helper.setUp();
    }

    @After
    public void tearDown() {
        helper.tearDown();
    }

    @Test
    public void createAndGetUserByXMPPID() {
        User u = new User();
        u.setXmppID("l@lds.li");
        u.insert();

        // now get the user
        User getUser = User.byXmppId("l@lds.li");
        assertNotNull(getUser);
        assertEquals(getUser.getXmppID(), u.getXmppID());

        // Makes sure we aren't leaking - will fail on subsequent runs
        DatastoreService ds = DatastoreServiceFactory.getDatastoreService();
        assertEquals(1, ds.prepare(new Query("User")).countEntities());
    }

    @Test
    public void userShouldNotBeRegisteredWihoutXMPPIDandOauth() {
        User u = new User();
        u.setXmppID("l@lds.li");
        u.insert();

        assertFalse(u.isActive());

        u.setTwitterToken("aa");
        u.setTwitterTokenSecret("xx");

        u.update();

        assertTrue(u.isActive());
    }

    @Test
    public void needsOauthReminder() {
        // if isn't active and hasn't been prompted in the last hour, re-prompt
        User u = new User();
        u.setXmppID("l@lds.li");
        u.insert();

        // we need to be reminded
        assertEquals(1, User.needsReminder().size());

        // set their last remind time to now.
        u.setLastReminded(new Date());
        u.update();

        // should no longer need reminder
        assertEquals(0, User.needsReminder().size());
    }

    @Test
    public void addRemoveWatchToExistingUser() {
        DatastoreService ds = DatastoreServiceFactory.getDatastoreService();
        assertEquals(0, ds.prepare(new Query("Watch")).countEntities());

        User u = new User();
        u.setXmppID("l@lds.li");

        u.insert();

        u.addWatch("testterm");

        assertEquals(1, ds.prepare(new Query("Watch")).countEntities());
        assertEquals("testterm", Watch.all(Watch.class).get().getTerm());

        u.removeWatch("testterm");

        assertEquals(0, ds.prepare(new Query("Watch")).countEntities());
    }

    @Test
    public void addRemoveWatchExisting() {
        DatastoreService ds = DatastoreServiceFactory.getDatastoreService();
        assertEquals(0, ds.prepare(new Query("Watch")).countEntities());

        User u1 = new User();
        u1.setXmppID("1@u.com");
        u1.insert();
        User u2 = new User();
        u2.setXmppID("2@u.com");
        u2.insert();

        u1.addWatch("term");
        assertEquals(1, ds.prepare(new Query("Watch")).countEntities());

        u2.addWatch("term");
        assertEquals(1, ds.prepare(new Query("Watch")).countEntities());

        u2.removeWatch("term");
        assertEquals(1, ds.prepare(new Query("Watch")).countEntities());

        u1.removeWatch("term");
        assertEquals(0, ds.prepare(new Query("Watch")).countEntities());
    }

}

File: test/UserTest.scala
import play.test._

import org.junit._
import org.scalatest.junit._
import org.scalatest._
import org.scalatest.matchers._
import com.google.appengine.tools.development.testing._
import com.google.appengine.api.datastore._

import models._

class UserTest extends UnitTest with ShouldMatchersForJUnit {
  var helper = new LocalServiceTestHelper(new LocalDatastoreServiceTestConfig)

  @Before def setUp = {
    helper.setUp
  }

  @After def tearDown = {
    helper.tearDown
  }

  @Test def createAndGetUserByXMPPID() {
    var u = new User()
    u.xmppID = "l@lds.li"
    u.save

    // now get the user
    var getUser = User.query.byXmppId("l@lds.li")
    
    getUser should not be (null)
    getUser.xmppID should equal (u.xmppID)

    // Makes sure we aren't leaking - will fail on subsequent runs
    var ds = DatastoreServiceFactory.getDatastoreService
    ds.prepare(new Query("User")).countEntities should equal (1)
  }


  @Test def userShouldNotBeRegisteredWihoutXMPPIDandOauth {
    var u = new User()
    u.xmppID = "l@lds.li"
    u.save

    u.isActive should be (false)

    u.twitterToken = "aa"
    u.twitterTokenSecret = "xx"

    u.save

    u.isActive should be (true)
  }

  @Test def needsOauthReminder {
    // if isn't active and hasn't been prompted in the last hour, re-prompt
    var u = new User()
    u.xmppID = "l@lds.li"
    u.save

    // we need to be reminded
    User.query.needsReminder().countAll should equal (1)

    // set their last remind time to now.
    u.lastReminded = new java.util.Date()
    u.save

    // should no longer need reminder
    User.query.needsReminder().countAll should equal (0)
  }

  /*@Test def addRemoveWatchToExistingUser {
    var ds = DatastoreServiceFactory.getDatastoreService
    ds.prepare(new Query("Watch")).countEntities should equal (0)

    var u = new User()
    u.xmppID = "l@lds.li"

    u.insert

    u.addWatch("testterm")

    ds.prepare(new Query("Watch")).countEntities should equal (1)
    Watch.all.get.getTerm should equal ("testterm")

    u.removeWatch("testterm")

    ds.prepare(new Query("Watch")).countEntities should equal (0)
  }

  @Test def addRemoveWatchExisting {
    var ds = DatastoreServiceFactory.getDatastoreService
    ds.prepare(new Query("Watch")).countEntities should equal (0)

    var u1 = new User()
    u1.xmppID = "1@u.com"
    u1.insert
    var u2 = new User()
    u2.xmppID = "2@u.com"
    u2.insert

    u1.addWatch("term")
    ds.prepare(new Query("Watch")).countEntities should equal (1)

    u2.addWatch("term")
    ds.prepare(new Query("Watch")).countEntities should equal (1)

    u2.removeWatch("term")
    ds.prepare(new Query("Watch")).countEntities should equal (1)

    u1.removeWatch("term");
    ds.prepare(new Query("Watch")).countEntities should equal (0)
  }*/
}
